src/_pytest/assertion/rewrite.py

```diff
@@ -402,12 +402,11 @@ def _saferepr(obj):
     JSON reprs.
 
     """
-    repr = py.io.saferepr(obj)
-    if isinstance(repr, six.text_type):
-        t = six.text_type
+    r = py.io.saferepr(obj)
+    if isinstance(r, six.text_type):
+        return r.replace(u"\n", u"\\n")
     else:
-        t = six.binary_type
-    return repr.replace(t("\n"), t("\\n"))
+        return r.replace(b"\n", b"\\n")
```

Comment:
Note that `bytes("string literal")` in python3+ is a `TypeError` so I fixed these as well.

As far as I can tell however this situation is impossible as `__repr__` raises for non `str` types in python3.x